pgen2/grammar.py

```diff
@@ -171,6 +171,72 @@ def dump_nonterminals_cpp(self, f):
           f.write("""\
 
 }  // namespace grammar_nt
+""")
+
+      def dump_cpp(self, header_f, src_f):
+          # type: (IO[str], IO[str]) -> None
+          """Dump a C++ implementation of the grammar tables."""
+          header_f.write("""\
+// This code is generated by pgen2/grammar.py
+# include "cpp/grammar.h"
+
+namespace grammar {
+
+extern grammar OilGrammar;
+
+}  // namespace grammar
+""")
+          src_f.write("""\
+// This code is generated by pgen2/grammar.py
+# include "cpp/grammar.h"
+
+namespace grammar {
+
+""")
+
+          def make_bitset(first):
+            # type: (Dict[int, int]) -> str
+            bits = ["0" for _ in range(256)]
+            for pos, isset in first.items():
+              bits[pos] = str(isset)
+
+            return ''.join(bits)
+
+          src_f.write('grammar OilGrammar = {\n')
+          src_f.write('  std::vector<dfa>{\n')
+          for i, (states, first) in self.dfas.items():
+            src_f.write('    dfa{\n')
+            src_f.write('        %d,\n' % i)
+            src_f.write('        "%s",\n' % self.number2symbol[i])
+            src_f.write('        0,\n')
+            src_f.write('        std::vector<state>{\n')
+            for arcs in states:
+              src_f.write('            state{\n')
+              src_f.write('                std::vector<arc>({\n')
+              for a, b in arcs:
+                src_f.write('                    {%d, %d},\n' % (a, b))
+
+              src_f.write('                }),\n')
+              src_f.write('                0, 0, nullptr, 0,\n')
+              src_f.write('            },\n')
+
+            src_f.write('        },\n')
+            src_f.write('        bitset_t{"%s"},\n' % make_bitset(first))
```

Comment:
`first_t` is basically a bitset, right? the corresponding field in cpython's grammar.h is

Reply:
I actually don't know but it seems very plausible :)